### DISCUSSION TITLE
Day Light Math with date-fns

### DISCUSSION THREAD

**Discussion by githubajay1mishra** [js-fns/date-fns#2615] [Q&A]
Our team is planning to replace moment with date-fns. We hit a road block while porting some functionality that worked correctly with moment. The issue is `addMonths` does not return the correct time when calculating next month time across day light savings.

Below is some sample code that shows difference with moment and date-fns.

date-fns addMonths returns a time that is an hour ealier.
`addMonths(new Date('2021-02-15T14:30:00.000Z'), 1).toISOString() // returns "2021-03-15T13:30:00.000Z"`  

Moment has considerations from day light saving  returns the correct time.
`moment(new Date('2021-02-15T14:30:00.000Z')).utcOffset(0).add(1, 'month').toDate().toISOString() //returns "2021-03-15T14:30:00.000Z `

How does date-fns handle daylight saving?
How should i invoke addMonths to get a similar behaviour like moment?

**Comment by catsclaw** [comment]
I've just run into the exact same problem. I've got a bunch of times representing midnight in UTC. I'm trying to add a day to them, which doesn't work if the day sits on a daylight savings boundary in the local time zone.

Basically, `date-fns` thinks "one day" = "24 hours" which simply _isn't_ true. It's hard to view this as anything but a bug. Even if you're working exclusively in local times, if you have a `Date` representing noon and you ask to add one day to it, you should get noon the next day. Sometimes you don't.

**Reply by tan75** [reply]
> I've just run into the exact same problem. I've got a bunch of times representing midnight in UTC. I'm trying to add a day to them, which doesn't work if the day sits on a daylight savings boundary in the local time zone.
> 
> Basically, `date-fns` thinks "one day" = "24 hours" which simply _isn't_ true. It's hard to view this as anything but a bug. Even if you're working exclusively in local times, if you have a `Date` representing noon and you ask to add one day to it, you should get noon the next day. Sometimes you don't.

Hi @catsclaw 
Could you please provide an example of the mentioned '24 hours day' behavior? 
This is simply not true and the result does depend upon a timezone and a date.

**Comment by tan75** [comment]
> Our team is planning to replace moment with date-fns. We hit a road block while porting some functionality that worked correctly with moment. The issue is `addMonths` does not return the correct time when calculating next month time across day light savings.
> 
> Below is some sample code that shows difference with moment and date-fns.
> 
> date-fns addMonths returns a time that is an hour ealier.
> `addMonths(new Date('2021-02-15T14:30:00.000Z'), 1).toISOString() // returns "2021-03-15T13:30:00.000Z"`
> 
> Moment has considerations from day light saving returns the correct time.
> `moment(new Date('2021-02-15T14:30:00.000Z')).utcOffset(0).add(1, 'month').toDate().toISOString() //returns "2021-03-15T14:30:00.000Z `
> 
> How does date-fns handle daylight saving?
> How should i invoke addMonths to get a similar behaviour like moment?

hi @githubajay1mishra  
What's your timezone?
I am getting a correct result here in Irish standard time.

![image](https://user-images.githubusercontent.com/8555279/134691875-43777336-9aa3-4792-8f35-fc90a6270a30.png)

**Reply by githubajay1mishra** [reply]
Hi @tan75 ,
I am in the 'America/New_York' timezone.

**Reply by githubajay1mishra** [reply]
Hey @tan75,
 Any updates on this? 
 Is daylight savings  supported with date-fns?
Or any documentation about handling daylight savings?

**Comment by kossnocorp** [comment]
date-fns don't have any particular logic for `addMonths` for DST handling. It works out of the box in modern runtimes. In some functions, we incorporate DST fixes that adjust [the time, like `addBusinessDays`](https://github.com/date-fns/date-fns/blob/a91e059dce796c580d54d5894dd9f19348dd86fa/src/addBusinessDays/index.ts#L64).

The thing is that you made a mistake in your examples. You show time in UTC, which correctly changes when New York goes into the daylight saving time.

![image](https://user-images.githubusercontent.com/52201/137257837-5e861937-d68d-43a6-921a-247c475b9b8c.png)

As you can see, when using `toString` that prints time in the local timezone, it shows the same hours.

**Reply by githubajay1mishra** [reply]
@kossnocorp I don't think it works out of the box correctly for `addMonths`
My input is a date time in UTC. Day light savings shouldn't impact my time calculations.

`const input = new Date('2021-02-15T14:30:00.000Z')`
`const expected = new Date('2021-03-15T14:30:00.000Z')`
`const actual = addMonths(input, 1)`
`\\ Returns -1 an hour earlier`
`console.log((actual.getTime() - expected.getTime()) / (1000 * 3600))`
